Habitica new user registration

juemura opened the pull request:
Missing: 
- update/delete user
- wrapper for other http requests
- Improve unit testing

AmplabJenkins commented:
Can one of the admins verify this patch?

shankari commented:
ok to test

shankari commented:
High level comment: in the future, please submit pull reviews from branches
Makes it easier to clean up bad mistakes and revert unnecessary changes

shankari commented on conf/net/keys.json:
```diff
@@ -10,5 +10,10 @@
+  },
+  "habitica" : {
+    "url": "http://54.159.38.241:3000",
+    "api_id": "Get from http://54.159.38.241:3000",
```
`api_id` and `api_token` are not used, so should be removed to avoid confusion

shankari commented on emission/net/ext_service/habitica/register.py:
```diff
@@ -0,0 +1,24 @@
+key_data = json.load(key_file)
+url = key_data["habitica"]["url"]
+
+def habiticaRegister(username, email, password, our_uuid):
```
@sunil07t and @juemura need to decide login flow. will the user specify a password or will we autogenerate on? Everybody hates coming up with new passwords every time and requiring a separate password means that the integration is not as seamless as we would like.

On the other hand, if we want to use the standard habitica, maybe we want to allow existing users to login to their existing account?

shankari commented on emission/tests/netTests/TestHabiticaRegister.py:
```diff
@@ -0,0 +1,39 @@
+  # so we don't test that here. Instead, we test the save and retrieval of the datastructure 
+  # that we get back from habitica. Some sample auth messages are shown above
+
+  # TODO: See if we can use Mock to create a mock habitica server that can allow
```
yes, this is a good suggestion

shankari commented on emission/tests/netTests/TestHabiticaRegister.py:
```diff
@@ -0,0 +1,39 @@
+    self.testUserUUID = uuid.uuid4()
+    self.sampleAuthMessage1 = {'username': '99999999999995543', 'email': '99999999999995543@test.com', 'password': 'aa!A15551', 'our_uuid': self.testUserUUID}
+
+  # One part of this is the communication with habitica. That is an integration test, not a unit test,
```
This is not accurate - this does actually test the communication with habitica since the register code makes a call to the server.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/e-mission-server-prb/565/
Test FAILed.

shankari commented:
I will merge this as soon as you fix the `conf` file and the tests pass.
Other comments are basically nits

shankari commented:
jenkins, test this please

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/e-mission-server-prb/566/
Test FAILed.

shankari commented:
Also, you need to include `sdk` in `requirements.txt` otherwise the import fails

```
======================================================================
ERROR: netTests.TestHabiticaRegister (unittest.loader.ModuleImportFailure)
----------------------------------------------------------------------
ImportError: Failed to import test module: netTests.TestHabiticaRegister
Traceback (most recent call last):
  File "/home/anaconda/lib/python2.7/unittest/loader.py", line 254, in _find_tests
    module = self._get_module_from_name(name)
  File "/home/anaconda/lib/python2.7/unittest/loader.py", line 232, in _get_module_from_name
    __import__(name)
  File "<https://amplab.cs.berkeley.edu/jenkins/job/e-mission-server-prb/ws/emission/tests/netTests/TestHabiticaRegister.py",> line 9, in <module>
    import emission.net.ext_service.habitica.register as reg
  File "emission/net/ext_service/habitica/register.py", line 7, in <module>
    from sdk import Habitica
ImportError: No module named sdk
```

Alternatively, since we are not using the sdk now, and it is not clear how much the sdk adds given the planned passthrough, we can just remove the sdk for now add it later when we really need it

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/e-mission-server-prb/571/
Test FAILed.

shankari commented:
Test failed because:

```
======================================================================
ERROR: testAddNewUser (netTests.TestHabiticaRegister.TestHabiticaRegister)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "<https://amplab.cs.berkeley.edu/jenkins/job/e-mission-server-prb/ws/emission/tests/netTests/TestHabiticaRegister.py",> line 22, in testAddNewUser
    sampleAuthMessage1Ad.password, sampleAuthMessage1Ad.our_uuid)
  File "emission/net/ext_service/habitica/register.py", line 26, in habiticaRegister
    'habitica_password': password, 'habitica_id': user_dict['data']['_id'],
KeyError: 'data'

----------------------------------------------------------------------
```

I think that we hit this yesterday - you get a 401 when you try to reuse the user email.
```
{u'message': u'Email address is already used in an account.', u'success': False, u'error': u'NotAuthorized'}
```

I guess you fixed it by changing the email address at that time. That doesn't work as more than a stopgap because the unit tests are run over and over again. If a particular parameter needs to be unique, it needs to be freshly generated each time.

Randomly generate username and user email in the test instead of hardcoding them?

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/e-mission-server-prb/573/
Test FAILed.

juemura commented:
Sorry, there was a silly mistake. Fixed now. I added code to randomly generate a string, but maybe you already have something like that and I can just import it? If not, this should work.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/e-mission-server-prb/574/
Test PASSed.

shankari commented:
> I added code to randomly generate a string, but maybe you already have something like that and I can just import it?

I don't think that I do. The only random generation that I do is the UUID generation. This looks good - we could move it into `common` if we need to use it other places as well. But for now, this is fine.

shankari commented:
Congratulations! First pull request merged?
